core/src/crds_value.rs

```diff
@@ -428,6 +428,23 @@ impl CrdsValueLabel {
             CrdsValueLabel::DuplicateShred(_, p) => *p,
         }
     }
+
+    /// Returns number of possible distinct labels of the same type for
+    /// a fixed pubkey, and None if that is practically unlimited.
+    pub(crate) fn value_space(&self) -> Option<usize> {
+        match self {
+            CrdsValueLabel::ContactInfo(_) => Some(1),
+            CrdsValueLabel::Vote(_, _) => Some(MAX_VOTES as usize),
+            CrdsValueLabel::LowestSlot(_) => Some(1),
+            CrdsValueLabel::SnapshotHashes(_) => Some(1),
+            CrdsValueLabel::EpochSlots(_, _) => Some(MAX_EPOCH_SLOTS as usize),
+            CrdsValueLabel::AccountsHashes(_) => Some(1),
+            CrdsValueLabel::LegacyVersion(_) => Some(1),
+            CrdsValueLabel::Version(_) => Some(1),
+            CrdsValueLabel::NodeInstance(_, _) => None,
+            CrdsValueLabel::DuplicateShred(_, _) => None,
```

Comment:
Can we not inherently limit these though? Can NodeInstance be limited to 2 unique values per pubkey?
And DuplicateShred could have a fixed limit, maybe around 50-100 shred entries per pubkey.

Reply:
With node-instance (which is causing trouble right now), it is very tricky. Because if there are 2 instances of the same node, and they pick the same index, then they will override each other's `NodeInstance` in the crds table, making it less likely for one to see the other (because every 3rd node can only have `NodeInstance` of one of these two in its crds table at each moment). That was the reason the token was added to the label in https://github.com/solana-labs/solana/pull/14037

With the two instances operating independently (and unaware of each other), it is not trivial how each can choose a different index not overriding the other's instance in the table.

Reply:
How would they pick the same index, isn't it random u64?

Reply:
The `token` is u64, and that is what it is included in the label, effectively making `NodeInstances` per pubkey unlimited:
https://github.com/solana-labs/solana/blob/30038a884/core/src/crds_value.rs#L395

What I was saying above is that it is not trivial to change:
```
NodeInstance(Pubkey, u64 /*token*/),
```
to
```
type NodeInstanceIndex = u8;
const MAX_NODE_INSTANCE_INDEX: NodeInstanceIndex = 2;

enum CrdsValueLabel {
  ... 
  NodeInstance(Pubkey, NodeInstanceIndex),
}
```
like what is done for `VoteIndex`:
https://github.com/solana-labs/solana/blob/30038a884/core/src/crds_value.rs#L28-L32

Because instances may pick the same `NodeInstanceIndex` overriding each others in gossip.

Reply:
Right, you would still have something like the custom logic above to clean it up because the label itself does not limit the number of values inherently. I guess what I'm saying instead of having this 512 limit, it can be even stricter for NodeInstance, just the two latest unique labels need to be present.

Reply:
> I guess what I'm saying instead of having this 512 limit, it can be even stricter for NodeInstance, just the two latest unique labels need to be present.

That is a fair point, but I tried to keep the logic here simple for now, to first stop the crds table unnecessarily getting bigger.

But, the actual bug here is that nodes keep generating new node-instances, which I am trying to debug why.

Reply:
Ok, yea that's fine. thanks.